### ISSUE TITLE
Integration with binder

### ISSUE THREAD

**Issue by jdblischak** [workflowr/workflowr#167]
[binder](https://binderhub.readthedocs.io/en/latest/) is an amazing tool for reproducible research. It would be great if we could make it easier for workflowr users to integrate binder into their research projects.

An idea for a function, e.g. `wflow_use_binder()`, would perform the following steps:

1. Search all the published R Markdown files for `library()` calls (and related pkg usage, e.g. `pkg::`)
1. Create and commit a [configuration file](https://repo2docker.readthedocs.io/en/latest/config_files.html) that is supported by [repo2docker](https://repo2docker.readthedocs.io).
1. Add a binder link to `README.md` and to the website (either `index.Rmd` or the navbar in `_site.yml`)

The first step is the weakest. I could imagine improving it in multiple ways. One would be to have each Rmd file automatically write its dependencies after running `sessionInfo()`. Alternatively, wait for the package manager functionality (#128) to be integrated before making this any more complicated.

I'm open to ideas and suggestions.

cc: @seabbs @sdhutchins 

xref: [Tweet](https://twitter.com/seabbs/status/1139518437917696003)

**Comment by pcarbo** [comment]
@jdblischak What do you think about [Code Ocean](https://codeocean.com)?

**Comment by jdblischak** [comment]
> What do you think about Code Ocean?

How so? In comparison to binder? To me, Code Ocean is an online platform for reproducible data science, similar to [The Whole Tale](https://wholetale.org/) and [Gigantum](https://docs.gigantum.com/). I'm not sure what meaningul features I could add to workflowr to integrate with these online platforms.

In contrast, Binder is built to work with GitHub repositories. It spins up a computing environment for any GitHub repo with a configuration file. Thus workflowr can assist with this setup.

Do you see a way that workflowr could better integrate with Code Ocean and similar online platforms?

**Comment by pcarbo** [comment]
@jdblischak See [here](https://help.codeocean.com/articles/2887105-version-control-on-code-ocean-an-overview) and [here](https://help.codeocean.com/articles/2887183-git-clone-your-capsule-to-your-computer-or-a-different-git-host): "Code Ocean capsule files are managed as a Git repository." So, for example, you could potentially use workflowr inside a Code Ocean "capsule".

I don't really know much about Code Ocean, only that is being pushed (and used) by many people. So I think it is something we should consider.

**Comment by jdblischak** [comment]
> So, for example, you could potentially use workflowr inside a Code Ocean "capsule".

Right, it looks like it should be possible to execute `wflow_build()`/`wflow_publish()` inside of a Code Ocean container.

> I don't really know much about Code Ocean, only that is being pushed (and used) by many people. So I think it is something we should consider.

I'm open to it. It's just unclear to me what feature(s) we can add.

Looking at this example project: https://codeocean.com/capsule/6314882/tree/v1, it appears that they are controlling the environment via a Dockerfile (and using conda to install software in the Docker image). Thus I guess one option would be to have more generic functions such as `wflow_use_docker()` and `wflow_use_conda()` that would output a configuration file based on the current Rmd files. Then a function like `wflow_use_binder()` would call one of these functions in addition to adding the binder badge. And a user interested in Code Ocean would only run `wflow_use_docker()` and then upload the project.

Though it seems that Code Ocean also has its own [package management system](https://help.codeocean.com/articles/1374271-configuring-your-computational-environment-an-overview). Also, I couldn't find an API for Code Ocean.

Do we know anyone that uses Code Ocean? That may be easier than trying to guess what a hypothetical Code Ocean user would want from workflowr.

**Comment by pcarbo** [comment]
> Do we know anyone that uses Code Ocean?

I don't, but I'm aiming to try it out.

**Comment by jdblischak** [comment]
> I don't, but I'm aiming to try it out.

OK. Please let me know what you think after giving it a try. My initial impression is that using workflowr on Code Ocean would be redundant. Code Ocean has the similar goal of abstracting away the difficulty of version control and making it easier to share results.

**Comment by jdblischak** [comment]
The new [holepunch](https://github.com/karthik/holepunch) package has a large overlap with most of the features I was considering developing:

> holepunch will read the contents of your R project on GitHub, create a DESCRIPTION file with all dependencies, write a Dockerfile, add a badge to your README, and build a Docker image.

And it is going to be released soon:

> The ETA for the first release of this package is early July.

Thus one option would be to wait for its release, and then integrate it into workflowr.

**Comment by seabbs** [comment]
If this works well then this could be a great option. Could include a wrapper function / add to the docs to point users towards this?

**Comment by jdblischak** [comment]
> Could include a wrapper function / add to the docs to point users towards this?

Agreed. Once it is released, at minimum we should add it to the docs. If you get the chance to test it out, please share your thoughts.

**Comment by jdblischak** [comment]
@timtrice has put together some instructions for integrating workflowr with binder via the holepunch package: [enhanced-workflowr](https://enhanced-workflowr.timtrice.net/index.html)

@timtrice Once holepunch is on CRAN, it would be great to have your help on adding a wrapper function to workflowr.

**Comment by jdblischak** [comment]
Quick update: holepunch still hasn't been released to CRAN.

There are two main strategies that holepunch provides for using binder:

1. `DESCRIPTION`+`Dockerfile` using `write_compendium_description()` and `write_dockerfile()`
1. `install.r`+`runtime.txt` using `write_install()` and `write_runtime()`

Both strategies should be compatible with a workflowr project, so we should provide an argument to allow users to choose which strategy to implement.
